src/optimizer/rules/unnesting_rules.cpp

```diff
@@ -105,4 +107,113 @@ void UnnestSingleJoinToInnerJoin::Transform(common::ManagedPointer<AbstractOptim
   transformed->emplace_back(std::move(output));
 }
 
+///////////////////////////////////////////////////////////////////////////////
+/// DependentSingleJoinToInnerJoin
+///////////////////////////////////////////////////////////////////////////////
+DependentSingleJoinToInnerJoin::DependentSingleJoinToInnerJoin() {
+  type_ = RuleType::DEPENDENT_JOIN_GET_TO_INNER_JOIN;
+
+  match_pattern_ = new Pattern(OpType::LOGICALSINGLEJOIN);
+  match_pattern_->AddChild(new Pattern(OpType::LEAF));
+  auto right_child = new Pattern(OpType::LOGICALFILTER);
+  auto agg_child = new Pattern(OpType::LOGICALAGGREGATEANDGROUPBY);
+  agg_child->AddChild(new Pattern(OpType::LEAF));
+  right_child->AddChild(agg_child);
+  match_pattern_->AddChild(right_child);
+}
+
+RulePromise DependentSingleJoinToInnerJoin::Promise(GroupExpression *group_expr) const {
+  return RulePromise::LOGICAL_PROMISE;
+}
+
+bool DependentSingleJoinToInnerJoin::Check(common::ManagedPointer<AbstractOptimizerNode> plan,
+                                           OptimizationContext *context) const {
+  (void)context;
+  (void)plan;
+
+  UNUSED_ATTRIBUTE auto children = plan->GetChildren();
+  TERRIER_ASSERT(children.size() == 2, "SingleJoin should have 2 children");
+  return true;
+}
+
+void DependentSingleJoinToInnerJoin::Transform(common::ManagedPointer<AbstractOptimizerNode> input,
+                                               std::vector<std::unique_ptr<AbstractOptimizerNode>> *transformed,
+                                               UNUSED_ATTRIBUTE OptimizationContext *context) const {
+  UNUSED_ATTRIBUTE auto single_join = input->Contents()->GetContentsAs<LogicalSingleJoin>();
+  TERRIER_ASSERT(single_join->GetJoinPredicates().empty(), "SingleJoin should have no predicates");
+  // From LOGICALSINGLEJOIN -> LOGICALFILTER -> LOGICALAGGREGATEANDGROUPBY
+  //  to LOGICALFILTER -> LOGICALINNERJOIN -> LOGICALFILTER -> LOGICALAGGREGATEANDGROUPBY
+  auto &memo = context->GetOptimizerContext()->GetMemo();
+  auto filter_expr = input->GetChildren()[1];
+  auto agg_expr = filter_expr->GetChildren()[0];
+  auto agg_group_id = agg_expr->GetChildren()[0]->Contents()->GetContentsAs<LeafOperator>()->GetOriginGroup();
+  const auto &agg_group_aliases_set = memo.GetGroupByID(agg_group_id)->GetTableAliases();
+  auto &filter_predicates = filter_expr->Contents()->GetContentsAs<LogicalFilter>()->GetPredicates();
+
+  std::vector<AnnotatedExpression> top_predicates;
+  std::vector<AnnotatedExpression> down_predicates;
```

Comment:
It might be clearer to refer to these as ancestor_predicates and descendant_predicates, respectively.